inspirehep/modules/workflows/tasks/refextract.py

```diff
@@ -184,3 +189,60 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
     return extract_references_from_text(
         raw_ref['value'], source=raw_ref['source'], custom_kbs_file=custom_kbs_file
     )
+
+
+def match_references(references):
+    """Match references to record ids given a list of references.
+
+    Args:
+        references: Schema compliant list of references
+
+    Returns:
+        associated_references: The same list with the associated record ids (if any)
+    """
+
+    for i, reference in enumerate(references):
+        matched_recid = match_reference(reference)
+        if matched_recid:
+            references[i]['record'] = get_record_ref(matched_recid, 'literature')
+
+    return references
+
+
+def match_reference(reference):
+    """Match references given a reference metadata using InspireMatcher queires.
+
+    Args:
+        reference: The reference metadata
+        config: The configurtaion(s) for InspireMatcher queries
+
+    Returns:
+        The record ID of the matched reference
+    """
+
+    config_default = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DEFAULT_CONFIG']
+    config_jcap_and_jhep = current_app.config['WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_CONFIG']
+    config_data = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DATA_CONFIG']
+
+    journal_title = get_value(reference, 'reference.publication_info.journal_title')
+    if journal_title in ['JCAP', 'JHEP']:
+        try:
+            if get_value(reference, 'reference.publication_info.year'):
+                reference['reference']['publication_info']['year'] = str(
+                    reference['reference']['publication_info']['year'])
+            result = next(match(reference, config_jcap_and_jhep))
+            return result['_source']['control_number']
+        except StopIteration:
+            pass
+
+    try:
+        result = next(match(reference, config_data))
+        return result['_source']['control_number']
+    except StopIteration:
+        pass
+
+    try:
+        result = next(match(reference, config_default))
+        return result['_source']['control_number']
+    except StopIteration:
+        pass
```

Comment:
Repeating the same code pattern 3 times (especially this close to each other) should make you think that a refactoring is possible: https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming).

Currently this function is an example of this other "Rule-of-three": https://en.wikipedia.org/wiki/Planescape#Rule-of-Three : )

Reply:
Will change.

Reply:
Another comment on this, while we can do what you suggested, I think this would lead to a bit too much abstraction. What do you think about that?

Reply:
Mh, I see what you mean, but I would still do it. Let's see what @turtle321 thinks about this?

Reply:
I woul probably try to avoid excetions for control flows if possible, thus I would have written:
```
result = next(match(reference, config_jcap_and_jhep), None)
if result:
    return result['_source']['control_number']
```